test/integration/multiplexed_upstream_integration_test.cc

```diff
@@ -640,4 +640,43 @@ TEST_P(MultiplexedUpstreamIntegrationTest, UpstreamGoaway) {
   cleanupUpstreamAndDownstream();
 }
 
+TEST_P(MultiplexedUpstreamIntegrationTest, EarlyDataRejected) {
+  if (!Runtime::runtimeFeatureEnabled(Runtime::conn_pool_new_stream_with_early_data_and_http3) ||
+      upstreamProtocol() != Http::CodecType::HTTP3) {
+    return;
+  }
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Kick off the initial request and make sure it's received upstream.
+  auto response = codec_client_->makeHeaderOnlyRequest(default_request_headers_);
+  waitForNextUpstreamRequest();
+
+  const Http::TestResponseHeaderMapImpl too_early_response_headers{{":status", "425"}};
+  upstream_request_->encodeHeaders(too_early_response_headers, true);
+  // 425 response will be retried by Envoy, so expect another upstream request.
```

Comment:
actually can we do this test for HTTP/2 and make sure that we don't retry on early data?

Reply:
included H2 in this test